pep-0684-per-interpreter-gil.rst

```diff
@@ -0,0 +1,642 @@
+
+This directly coincides with an ongoing effort of many years to greatly
+reduce internal use of C global variables and consolidate the runtime
+state into ``_PyRuntimeState`` and ``PyInterpreterState``.
+(See `Consolidating Runtime Global State`_ below.)  That project has
+`significant merit on its own <Benefits to Consolidation_>`_
+and has faced little controversy.  So, while a per-interpreter GIL
+relies on the completion of that effort, that project should not be
+considered a part of this proposal, only a dependency.
+
+Other Isolation Considerations
+------------------------------
+
+CPython's interpreters must be strictly isolated from each other, with
+few exceptions.  To a large extent they already are.  Each interpreter
+has its own copy of all modules, classes, functions, and variables.
+The CPython C-API docs `explain further <caveats_>`_.
+
+.. _caveats: https://docs.python.org/3/c-api/init.html#bugs-and-caveats
+
+However, aside from what has already been mentioned (e.g. the GIL),
+there are a couple of ways in which interpreters still share some state.
+
+First of all, some process-global resources (e.g. memory,
+file descriptors, environment variables) are shared.  There are no
+plans to change this.
+
+Second, some isolation is faulty due to bugs or implementations that
+did not take multiple interpreters into account.  This includes
+CPython's runtime and the stdlib, as well as extension modules that
+rely on globals variables.  In these cases bugs should be opened.
+(Some are already.)
+
+Depending on Immortal Objects
+-----------------------------
+
+:pep:`683` introduces immortal objects as a CPython-internal feature.
+With immortal objects we can share any otherwise immutable global
+objects between all interpreters.  Consequently, this PEP does not
+need to address how to deal with the various objects
+`exposed in the public C-API <capi objects_>`_.
+It also simplifies the question of what to do about the builtin
+static types.  (See `Global Objects`_ below.)
+
+Both matters have alternate solutions but everything is simpler with
+immortal objects.  If PEP 683 is not accepted then this one will be
+updated with the alternatives.  This lets us reduce noise in this
+proposal.
+
+
+Motivation
+==========
+
+The fundamental problem we're solving here is a lack of true multi-core
+parallelism (for Python code) in the CPython runtime.  The GIL is the
+cause.  While it usually isn't a problem in practice, at the very least
+it makes Python's multi-core story murky, which makes the GIL
+a consistent distraction.
+
+Isolated interpreters are also an effective mechanism to support
+certain concurrency models.  :pep:`554` discusses this in more detail.
+
+Indirect Benefits
+-----------------
+
+Most of the effort needed for a per-interpreter GIL has benefits that
+make those tasks worth doing anyway:
+
+* makes multiple-interpreter behavior more reliable
+* has lead to fixes for long-standing runtime bugs that otherwise
```

Comment:
```suggestion
* has led to fixes for long-standing runtime bugs that otherwise
```

Not essential, but I'd be curious what bugs this is referring to.